# Bouncy Castle FAQ – RAG Pipeline Exploration

This notebook walks through the completed ingest–search pipeline step by step.
It demonstrates every function from tasks 1–9 without running the test suite.

## 1. Setup

Import all public functions from the project's `src` package and create a working directory (`db/`) for indexes and the DuckDB database.

In [1]:
import json
import os
import pathlib
import pickle

import duckdb
import faiss

from src.faqs import load_faqs
from src.pipeline import run_pipeline
from src.ingest import build_indexes
from src.search import search
from dotenv import load_dotenv

load_dotenv()

db_dir = pathlib.Path("db")
db_dir.mkdir(exist_ok=True)
print(f"Working directory: {db_dir.resolve()}")

/Users/henrik/Documents/dev/python/llm-zoomcamp/llm-zoomcamp-rag/.venv/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Working directory: /Users/henrik/Documents/dev/python/llm-zoomcamp/llm-zoomcamp-rag/db


## 2. Load FAQ Data

Use `load_faqs()` to read the CSV, verify row count (42), inspect a few sample rows, and check for missing values.

In [2]:
faqs = load_faqs()
print(f"Rows: {len(faqs)}")
print(f"Columns: {', '.join(faqs[0].keys())}")
print()
print("--- 3 Sample Rows ---")
for row in faqs[:3]:
    print(f"  [{row['Category']}] {row['Question']}")
    print(f"  Answer: {row['Answer'][:80]}...")
    print()

missing = {k: sum(1 for row in faqs if not row.get(k)) for k in faqs[0]}
print("--- Missing Values per Column ---")
for col, count in missing.items():
    print(f"  {col}: {count} missing")

Rows: 41
Columns: Category, Question, Answer

--- 3 Sample Rows ---
  [Booking & Reservations] How far in advance should I book?
  Answer: It's recommended to book as early as possible. Many companies accept bookings up...

  [Booking & Reservations] Is a deposit required to reserve a bouncy castle?
  Answer: Yes. Most companies require a non-refundable deposit (typically €50 or 10-50% of...

  [Booking & Reservations] What payment methods are accepted?
  Answer: Most rental companies accept credit cards. bank transfers. cash. and PayPal. Som...

--- Missing Values per Column ---
  Category: 0 missing
  Question: 0 missing
  Answer: 0 missing


## 3. dlt Pipeline

Run the dlt pipeline into a local DuckDB file inside `db/`. Query the loaded table and confirm idempotency by re-running the pipeline.

In [3]:
db_path = db_dir / "test_faq.duckdb"
pipeline, info = run_pipeline(
    destination="duckdb",
    dataset_name="faq",
    pipeline_name="test_faq_pipeline",
    db_path=db_path,
)
print("First run:")
print(info)

con = duckdb.connect(str(db_path))
count = con.sql('SELECT COUNT(*) FROM "faq"."faq_resource"').fetchone()[0]
print(f"Rows loaded: {count}")
print()

categories = con.sql('SELECT DISTINCT Category FROM "faq"."faq_resource" ORDER BY Category').fetchall()
print("Categories in the dataset:")
for (cat,) in categories:
    print(f"  - {cat}")
con.close()

First run:
Pipeline test_faq_pipeline load step completed in 0.04 seconds
1 load package(s) were loaded to destination duckdb and into dataset faq
The duckdb destination used duckdb:////Users/henrik/Documents/dev/python/llm-zoomcamp/llm-zoomcamp-rag/db/test_faq.duckdb location to store data
Load package 1785495792.866388 is LOADED and contains no failed jobs
Rows loaded: 41

Categories in the dataset:
  - Booking & Reservations
  - Cancellation & Weather Policies
  - Cleaning & Damage
  - Delivery Setup & Pickup
  - General Questions
  - Insurance & Liability
  - Location-Specific Questions
  - Pricing & Additional Fees
  - Safety Rules & Supervision
  - Setup Requirements


In [4]:
_pipeline2, _info2 = run_pipeline(
    destination="duckdb",
    dataset_name="faq",
    pipeline_name="test_faq_pipeline",
    db_path=db_path,
)
con2 = duckdb.connect(str(db_path))
count2 = con2.sql('SELECT COUNT(*) FROM "faq"."faq_resource"').fetchone()[0]
con2.close()

print(f"Rows after re-run: {count2}")
print(f"Idempotent (row count unchanged): {count2 == count}")

Rows after re-run: 41
Idempotent (row count unchanged): True


## 4. Build Indexes

Build BM25 and FAISS indexes from the FAQ data, saving them into `db/`. Then inspect each index to confirm structure.

In [5]:
bm25_path = db_dir / "bm25_index.pkl"
faiss_path = db_dir / "faiss_index.bin"
docs_path = db_dir / "ingest_docs.json"

result = build_indexes(
    faqs=faqs,
    bm25_path=bm25_path,
    faiss_path=faiss_path,
    docs_path=docs_path,
    force=True,
)
print("Index files created:")
for k, v in result.items():
    print(f"  {k}: {v}")

Loading weights: 100%|██████████| 103/103 [00:00<00:00, 11922.54it/s]


Index files created:
  bm25_path: db/bm25_index.pkl
  faiss_path: db/faiss_index.bin
  docs_path: db/ingest_docs.json


In [6]:
with open(bm25_path, "rb") as f:
    bm25 = pickle.load(f)
print(f"BM25 corpus size: {bm25.corpus_size} documents")
print(f"BM25 vocabulary size: {len(bm25.idf)} tokens")
print()

index = faiss.read_index(str(faiss_path))
print(f"FAISS index ntotal: {index.ntotal} vectors")
print(f"FAISS index dimension: {index.d}")
print()

with open(docs_path, encoding="utf-8") as f:
    docs_data = json.load(f)
print(f"Docs JSON entries: {len(docs_data)}")
print(f"First doc keys: {list(docs_data[0].keys())}")
print(f"Sample doc: {json.dumps(docs_data[0], ensure_ascii=False)}")

BM25 corpus size: 41 documents
BM25 vocabulary size: 516 tokens

FAISS index ntotal: 41 vectors
FAISS index dimension: 384

Docs JSON entries: 41
First doc keys: ['id', 'text', 'category', 'question', 'answer']
Sample doc: {"id": "faq_0", "text": "Booking & Reservations: How far in advance should I book? It's recommended to book as early as possible. Many companies accept bookings up to a year in advance with a deposit. Popular dates fill up quickly.", "category": "Booking & Reservations", "question": "How far in advance should I book?", "answer": "It's recommended to book as early as possible. Many companies accept bookings up to a year in advance with a deposit. Popular dates fill up quickly."}


## 5. Hybrid Search

Search using BM25 + FAISS with Reciprocal Rank Fusion (RRF). Try different queries, compare `k=1` vs `k=5`, and test the empty-query edge case.

In [7]:
def show_results(query, results):
    print(f"Query: {query!r}")
    print(f"Results: {len(results)}")
    for i, r in enumerate(results, 1):
        print(f"  {i}. [{r['category']}] {r['question']}")
        print(f"     Score: {r['score']}  |  Answer: {r['answer'][:60]}...")
    print()

In [8]:
results_booking = search("booking", k=5, bm25_path=bm25_path, faiss_path=faiss_path, docs_path=docs_path)
show_results("booking", results_booking)

Loading weights: 100%|██████████| 103/103 [00:00<00:00, 14458.28it/s]


Query: 'booking'
Results: 5
  1. [Booking & Reservations] Do you have a referral program?
     Score: 0.0333  |  Answer: Some companies offer referral credits (e.g. €25 per referral...
  2. [Booking & Reservations] What payment methods are accepted?
     Score: 0.0328  |  Answer: Most rental companies accept credit cards. bank transfers. c...
  3. [Booking & Reservations] Do you offer discounts for multiple-day rentals?
     Score: 0.032  |  Answer: Yes. many companies offer discounted rates for multi-day ren...
  4. [Booking & Reservations] How far in advance should I book?
     Score: 0.032  |  Answer: It's recommended to book as early as possible. Many companie...
  5. [Setup Requirements] What if my setup location is inaccessible or unsafe?
     Score: 0.0156  |  Answer: If the delivery team cannot safely set up due to inaccessibl...



In [9]:
results_cost = search("cost", k=5, bm25_path=bm25_path, faiss_path=faiss_path, docs_path=docs_path)
show_results("cost", results_cost)

Query: 'cost'
Results: 5
  1. [Pricing & Additional Fees] How much does it cost to rent a bouncy castle?
     Score: 0.0325  |  Answer: Prices vary by size. type. and location: Small basic bounce ...
  2. [Setup Requirements] What type of surface can the bouncy castle be set up on?
     Score: 0.0167  |  Answer: Most companies can set up on grass. concrete. asphalt. or in...
  3. [Pricing & Additional Fees] Are there any hidden fees I should know about?
     Score: 0.0167  |  Answer: Common additional charges: delivery fees for out-of-area loc...
  4. [Pricing & Additional Fees] What is included in the rental price?
     Score: 0.0164  |  Answer: Most rental prices include the inflatable unit. delivery. se...
  5. [General Questions] Can I use a water slide if it's not hot outside?
     Score: 0.0161  |  Answer: Water slides can be used as dry slides if the weather is coo...



In [10]:
results_safety = search("safety", k=5, bm25_path=bm25_path, faiss_path=faiss_path, docs_path=docs_path)
show_results("safety", results_safety)

Query: 'safety'
Results: 5
  1. [Safety Rules & Supervision] Is adult supervision required?
     Score: 0.0328  |  Answer: Yes. adult supervision (age 18+) is required at all times wh...
  2. [Safety Rules & Supervision] Why is silly string prohibited?
     Score: 0.0315  |  Answer: Silly string is highly flammable and can damage the inflatab...
  3. [Insurance & Liability] Am I responsible for injuries or damage?
     Score: 0.0167  |  Answer: The renter assumes responsibility for the safety of all user...
  4. [Safety Rules & Supervision] Do children need to wear socks?
     Score: 0.0164  |  Answer: Yes. socks are typically required to protect children's feet...
  5. [Safety Rules & Supervision] What items are prohibited in the bouncy castle?
     Score: 0.0164  |  Answer: Face paints. glitter. silly string. party poppers. confetti....



In [11]:
print("--- k=1 vs k=5 ---")
r1 = search("deposit", k=1, bm25_path=bm25_path, faiss_path=faiss_path, docs_path=docs_path)
print(f"k=1 returned {len(r1)} result{'s' if len(r1)!=1 else ''}: {r1[0]['question'] if r1 else '(none)'}")
print()
r5 = search("deposit", k=5, bm25_path=bm25_path, faiss_path=faiss_path, docs_path=docs_path)
print(f"k=5 returned {len(r5)} results:")
for r in r5:
    print(f"  - {r['question']}  (score: {r['score']})")

--- k=1 vs k=5 ---
k=1 returned 1 result: Is a deposit required to reserve a bouncy castle?

k=5 returned 5 results:
  - Is a deposit required to reserve a bouncy castle?  (score: 0.0333)
  - What is the cancellation policy?  (score: 0.0325)
  - How far in advance should I book?  (score: 0.032)
  - What payment methods are accepted?  (score: 0.0164)
  - What happens if it rains or there's bad weather?  (score: 0.0159)


In [12]:
print("--- Empty Query Edge Case ---")
empty_results = search("", k=5, bm25_path=bm25_path, faiss_path=faiss_path, docs_path=docs_path)
print(f"Empty query returned {len(empty_results)} result{'s' if len(empty_results)!=1 else ''}")
if empty_results:
    for r in empty_results:
        print(f"  - {r['question']}  (score: {r['score']})")
else:
    print("  (no results — empty query handled gracefully)")

--- Empty Query Edge Case ---
Empty query returned 5 results
  - Can I use a water slide if it's not hot outside?  (score: 0.0167)
  - What is included in the rental price?  (score: 0.0167)
  - What are the capacity limits?  (score: 0.0164)
  - Who sets up the bouncy castle?  (score: 0.0164)
  - Can I keep the bouncy castle overnight?  (score: 0.0161)


## 6. LLM Client

The `src.llm` module wraps Groq as the primary LLM provider with OpenAI as a fallback.
It handles rate limiting, error recovery, and returns structured metadata.

Key function:
- `ask_llm(system_prompt, user_message, groq_model=None, openai_model=None)` — calls Groq
  first; falls back to OpenAI on failure. Returns provider, model, latency, token counts.

Rate limiting for Groq's free tier:
- `GROQ_RPM_LIMIT`: 25 requests per minute
- `GROQ_RPD_LIMIT`: 900 requests per day

When the per-minute limit is reached, `_enforce_groq_rate_limits()` sleeps until a slot opens.
If the per-day limit is reached, a `RuntimeError` is raised.

In [13]:
import os

from src.llm import (
    ask_llm,
    DEFAULT_GROQ_MODEL,
    DEFAULT_OPENAI_MODEL,
    GROQ_RPM_LIMIT,
    GROQ_RPD_LIMIT,
)

print(f"Default Groq model: {DEFAULT_GROQ_MODEL}")
print(f"Default OpenAI model: {DEFAULT_OPENAI_MODEL}")
print(f"Groq RPM limit: {GROQ_RPM_LIMIT}")
print(f"Groq RPD limit: {GROQ_RPD_LIMIT}")

Default Groq model: llama-3.3-70b-versatile
Default OpenAI model: gpt-5.4-mini
Groq RPM limit: 25
Groq RPD limit: 900


### Real LLM Call

Below we call `ask_llm()` with a FAQ-related question.
If `GROQ_API_KEY` (or `OPENAI_API_KEY` as fallback) is set, a real LLM response is returned.
Otherwise the call is skipped with an explanatory message.

In [14]:
import os
from src.llm import ask_llm
system_prompt = (
    "You are a helpful assistant for a bouncy castle rental company. "
    "Answer the user's question based on the FAQ context provided."
)
question = "What happens if it rains on the day of my rental?"

groq_key = os.environ.get("GROQ_API_KEY")
openai_key = os.environ.get("OPENAI_API_KEY")

if groq_key or openai_key:
    result = ask_llm(system_prompt, question)
    print(f"Provider: {result['provider']}")
    print(f"Model: {result['model']}")
    print(f"Latency: {result['latency']}s")
    print(f"Tokens: {result['tokens']}")
    print()
    print("Response:")
    print(result['response'])
else:
    print("No API keys found in environment. No real LLM call was made.")
    print("Set GROQ_API_KEY (or OPENAI_API_KEY as fallback) to see a live response.")

Provider: openai
Model: gpt-5.4-mini
Latency: 1.63s
Tokens: {'prompt': 46, 'completion': 47, 'total': 93}

Response:
If the weather looks bad or rain is expected, we may need to reschedule or adjust the rental for safety. Please contact us as early as possible so we can review your options and help with the best next step.


## 7. RAG Pipeline

The `src.rag` module wires retrieval + LLM call into a single `answer_question()` function.
It uses the indexes built in section 4 to retrieve relevant FAQ entries, then passes them
as context to the LLM for answer generation.

In [15]:
from src.rag import answer_question
print("answer_question imported successfully")

answer_question imported successfully


### Real RAG Call

Below we call `answer_question()` with a sample question.
If `GROQ_API_KEY` (or `OPENAI_API_KEY` as fallback) is set, a real LLM response is returned.
Otherwise the call is skipped with an explanatory message.

In [16]:
groq_key = os.environ.get("GROQ_API_KEY")
openai_key = os.environ.get("OPENAI_API_KEY")

if groq_key or openai_key:
    result = answer_question(
        "What's your cancellation policy?",
        k=5,
        bm25_path=bm25_path,
        faiss_path=faiss_path,
        docs_path=docs_path,
    )
    print("Answer:")
    print(result["answer"])
    print()
    print(f"Provider: {result['provider']}  |  Model: {result['model']}")
    print(f"Latency: {result['latency']}s  |  Tokens: {result['tokens']}")
    print()
    print("--- Retrieved Contexts ---")
    for i, ctx in enumerate(result['contexts'], 1):
        print(f"  {i}. [{ctx['category']}] {ctx['question']}  (score: {ctx['score']})")
else:
    print("No API keys found in environment. No real LLM call was made.")
    print("Set GROQ_API_KEY (or OPENAI_API_KEY as fallback) to see a live response.")

Answer:
Cancellation policy depends on how far in advance you cancel:

- **14+ days notice:** Full refund minus processing fees, or your deposit can be held for rescheduling.
- **5–13 days notice:** Your deposit may be forfeited.
- **Less than 48 hours notice:** Typically a **50% cancellation fee** or **full deposit forfeiture**.

If you want, I can also share the weather-related cancellation policy.

Provider: openai  |  Model: gpt-5.4-mini
Latency: 1.397s  |  Tokens: {'prompt': 348, 'completion': 91, 'total': 439}

--- Retrieved Contexts ---
  1. [Cancellation & Weather Policies] What is the cancellation policy?  (score: 0.0333)
  2. [Cancellation & Weather Policies] What happens if it rains or there's bad weather?  (score: 0.0325)
  3. [Cancellation & Weather Policies] When can I cancel due to weather?  (score: 0.0325)
  4. [General Questions] Can I use a water slide if it's not hot outside?  (score: 0.0159)
  5. [General Questions] What if I need to extend my rental time on the day

### Empty-Context Edge Case

When the search results do not contain relevant information, the LLM
prompt explicitly states that no FAQ entries were found. The LLM should
respond that it doesn't have enough information.

In [17]:
if groq_key or openai_key:
    result = answer_question(
        "Can I use a bouncy castle on the moon?",
        k=5,
        bm25_path=bm25_path,
        faiss_path=faiss_path,
        docs_path=docs_path,
    )
    print("Answer:")
    print(result["answer"])
    print()
    print(f"Provider: {result['provider']}  |  Model: {result['model']}")
    print(f"Latency: {result['latency']}s  |  Tokens: {result['tokens']}")
else:
    print("No API keys found -- skipping edge case demonstration.")
    print("Set GROQ_API_KEY (or OPENAI_API_KEY as fallback) to see the LLM gracefully handle empty context.")

Answer:
The FAQ doesn’t cover using a bouncy castle on the moon, so I can’t answer that from the provided information.

Provider: openai  |  Model: gpt-5.4-mini
Latency: 1.042s  |  Tokens: {'prompt': 366, 'completion': 28, 'total': 394}


## 8. Postgres Logging Layer

The `src.db` module provides a Postgres-backed logging layer for RAG interactions.
It stores every question, answer, feedback, and metadata in a `rag_logs` table.

Key functions:
- `init_db()` — creates the `rag_logs` table if it doesn't exist
- `log_interaction(question, answer, metadata)` — logs a RAG interaction, returns the row ID
- `update_feedback(interaction_id, feedback)` — sets thumbs-up/down feedback on a logged interaction

The table schema uses JSONB for flexible metadata storage and SERIAL for auto-incrementing IDs.

> **Prerequisite:** [Docker Desktop](https://www.docker.com/products/docker-desktop/) must be running so Postgres is available.
> Start it with `docker compose up -d postgres` from the project root.

In [18]:
from src.db import init_db, log_interaction, update_feedback, get_connection, CREATE_TABLE_SQL

print("=== rag_logs Table Schema ===")
print(CREATE_TABLE_SQL)

=== rag_logs Table Schema ===

CREATE TABLE IF NOT EXISTS rag_logs (
    id SERIAL PRIMARY KEY,
    question TEXT NOT NULL,
    answer TEXT NOT NULL,
    feedback TEXT,
    metadata JSONB,
    created_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
    updated_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
);



In [19]:
import os

DATABASE_URL = os.environ.get("DATABASE_URL")

if DATABASE_URL:
    try:
        conn = get_connection(dsn=DATABASE_URL)
        init_db(conn=conn)

        log_id = log_interaction(
            "What's your cancellation policy?",
            "You can cancel up to 24 hours before your rental.",
            metadata={
                "provider": "groq",
                "model": "llama-3.3-70b-versatile",
                "tokens": {"prompt": 100, "completion": 50, "total": 150},
            },
            conn=conn,
        )
        print(f"Logged interaction with ID: {log_id}")

        update_feedback(log_id, "helpful", conn=conn)
        print(f"Feedback updated for ID {log_id}")

        with conn.cursor() as cur:
            cur.execute("SELECT id, question, feedback FROM rag_logs WHERE id = %s", (log_id,))
            row = cur.fetchone()
            print(f"Verified: id={row[0]}, question={row[1]!r}, feedback={row[2]!r}")

        conn.close()
        print("\nPostgres logging demonstration completed successfully.")
    except Exception as e:
        print(f"Could not connect to Postgres: {e}")
        print("Make sure DATABASE_URL points to a running Postgres instance.")
else:
    print("No DATABASE_URL found in environment.")
    print("Set DATABASE_URL to see a live Postgres logging demonstration.")
    print("Example: DATABASE_URL=postgres://user:pass@localhost:5432/rag_logs")

Logged interaction with ID: 123
Feedback updated for ID 123
Verified: id=123, question="What's your cancellation policy?", feedback='helpful'

Postgres logging demonstration completed successfully.


## 9. Evaluation: retrieval metrics

This section evaluates **retrieval quality only** — how well `search()` finds
the right FAQ documents (hit rate@k and MRR@k). LLM answer quality
evaluation is a separate topic (see tasks.md §10).

The `src.evaluate` module computes retrieval metrics against a ground-truth
dataset. Ground truth questions are in `data/ground_truth.csv` — 205
query–document pairs generated from the 41 FAQs using OpenAI (see
`generate_ground_truth.py`).

Note: Retrieval evaluation runs locally on pre-built indexes so it incurs
**zero LLM cost** — the cost estimates in section 10 are separate.

To regenerate ground truth yourself:

```bash
uv run python generate_ground_truth.py
```

Requires `OPENAI_API_KEY` in `.env`. This step is fully optional.

In [20]:
from src.evaluate import load_ground_truth, compute_hit_rate, compute_mrr, evaluate_retrieval

from src.faqs import load_faqs
from src.evaluate import load_ground_truth, compute_hit_rate, compute_mrr, evaluate_retrieval

faqs = load_faqs()
faq_by_id = {f"faq_{i}": row["Question"] for i, row in enumerate(faqs)}

gt = load_ground_truth()
print(f"Ground truth entries: {len(gt)}")
print()
by_doc = {}
for item in gt:
    by_doc.setdefault(item['document_id'], []).append(item['question'])
print(f"Unique FAQ IDs: {len(by_doc)}")
print()
for doc_id in sorted(by_doc):
    questions = by_doc[doc_id]
    original_q = faq_by_id.get(doc_id, "?")
    print(f"{doc_id} ({len(questions)} queries) — original: {original_q}")
    for q in questions:
        print(f"  - {q}")

Ground truth entries: 205

Unique FAQ IDs: 41

faq_0 (5 queries) — original: How far in advance should I book?
  - How early should I book a bounce house rental?
  - Can I reserve a bouncy castle months ahead?
  - How far ahead do you take bookings?
  - Do I need to book early for popular dates?
  - Can I book a party inflatable up to a year in advance?
faq_1 (5 queries) — original: Is a deposit required to reserve a bouncy castle?
  - Do I need to pay a deposit to book a bouncy castle?
  - Is there a deposit required when reserving a bounce house?
  - How much deposit do you need for a bouncy castle booking?
  - Can I secure a bouncy castle without paying upfront?
  - Is the booking deposit refundable for a bouncy castle?
faq_10 (5 queries) — original: What type of surface can the bouncy castle be set up on?
  - Can you put a bouncy castle on grass or concrete?
  - What surfaces can an inflatable castle be set up on?
  - Do you need grass for a bounce house, or can it go on asphalt/in

In [21]:
from src.search import search

sample = gt[10]
results = search(sample["question"], k=5, bm25_path=bm25_path, faiss_path=faiss_path, docs_path=docs_path)
target_id = sample["document_id"]

print(f"Query: {sample['question']}")
print(f"Target document: {target_id}")
print(f"Search results:")
for i, r in enumerate(results, 1):
    print(f"  {i}. {r['id']} — {r['question'][:60]}")
print()
print(f"Hit rate@{len(results)}: {compute_hit_rate(results, [target_id])}")
print(f"MRR:                  {compute_mrr(results, [target_id])}")

Query: How early should I book a bounce house rental?
Target document: faq_0
Search results:
  1. faq_0 — How far in advance should I book?
  2. faq_19 — What are the capacity limits?
  3. faq_36 — How much does it cost to rent a bouncy castle?
  4. faq_14 — What time will delivery and pickup occur?
  5. faq_20 — Are there age or height restrictions?

Hit rate@5: 1.0
MRR:                  1.0


In [22]:
report = evaluate_retrieval(k=5, bm25_path=bm25_path, faiss_path=faiss_path, docs_path=docs_path)
print(f"Hit Rate@{report['k']}: {report['hit_rate']}")
print(f"MRR@{report['k']}:      {report['mrr']}")
print(f"Evaluated on {report['n']} queries")
print()
print("--- First 10 Per-Query Details ---")
for d in report['details'][:10]:
    hit_mark = "✓" if d['hit'] else "✗"
    print(f"  {hit_mark} {d['question'][:50]:50s}  mrr={d['mrr']:.4f}  doc={d['document_id']}")

Hit Rate@5: 0.9463
MRR@5:      0.8211
Evaluated on 205 queries

--- First 10 Per-Query Details ---
  ✓ What payment options do you accept for renting a b  mrr=0.5000  doc=faq_2
  ✓ Can I pay with credit card, cash, or PayPal?        mrr=1.0000  doc=faq_2
  ✓ Do bounce house rentals take bank transfers too?    mrr=1.0000  doc=faq_2
  ✓ Which payment methods are available when booking a  mrr=1.0000  doc=faq_2
  ✓ Are there any fees for returned payments or bounce  mrr=1.0000  doc=faq_2
  ✓ What’s your cancellation policy for a bounce house  mrr=1.0000  doc=faq_5
  ✗ How much do I get back if I cancel my bouncy castl  mrr=0.0000  doc=faq_5
  ✓ If I cancel 2 weeks before, do I get a full refund  mrr=0.5000  doc=faq_5
  ✓ What happens to my deposit if I cancel 5-13 days b  mrr=1.0000  doc=faq_5
  ✓ How bad is the cancellation fee if I cancel less t  mrr=1.0000  doc=faq_5


## 10. Keyword Search Evaluation

Evaluate a DuckDB-backed keyword (ILIKE) search alongside the hybrid BM25+FAISS search.
The `keyword_search()` function runs SQL ILIKE on `faq.faq_resource` across
`Question`, `Answer`, and `Category` with tunable per-field weights.

`evaluate_retrieval()` accepts a pluggable `search_fn=` parameter so the same metric
computation can be reused for any search backend.

In [23]:
from src.search import keyword_search, search
from src.evaluate import evaluate_retrieval, load_ground_truth

gt = load_ground_truth()

kw_report = evaluate_retrieval(
    ground_truth=gt,
    k=5,
    search_fn=keyword_search,
    db_path=db_path,
    docs_path=docs_path,
)
print(f"Keyword ILIKE — Hit Rate@{kw_report['k']}: {kw_report['hit_rate']}")
print(f"Keyword ILIKE — MRR@{kw_report['k']}:      {kw_report['mrr']}")
print(f"Evaluated on {kw_report['n']} queries")

Keyword ILIKE — Hit Rate@5: 0.7561
Keyword ILIKE — MRR@5:      0.5821
Evaluated on 205 queries


### Hybrid vs Keyword Comparison

Compare the two search strategies side by side.

In [24]:
hybrid_report = evaluate_retrieval(
    ground_truth=gt,
    k=5,
    bm25_path=bm25_path,
    faiss_path=faiss_path,
    docs_path=docs_path,
)

print(f"{'Strategy':<20} {'Hit Rate@5':>12} {'MRR@5':>12} {'n':>6}")
print(f"{'--------':<20} {'----------':>12} {'------':>12} {'---':>6}")
print(f"{'Hybrid BM25+FAISS':<20} {hybrid_report['hit_rate']:>12.4f} {hybrid_report['mrr']:>12.4f} {hybrid_report['n']:>6}")
print(f"{'Keyword ILIKE':<20} {kw_report['hit_rate']:>12.4f} {kw_report['mrr']:>12.4f} {kw_report['n']:>6}")
print()

delta_hr = kw_report['hit_rate'] - hybrid_report['hit_rate']
delta_mrr = kw_report['mrr'] - hybrid_report['mrr']
print(f"Keyword vs Hybrid — Hit Rate delta: {delta_hr:+.4f}")
print(f"Keyword vs Hybrid — MRR delta:      {delta_mrr:+.4f}")

Strategy               Hit Rate@5        MRR@5      n
--------               ----------       ------    ---
Hybrid BM25+FAISS          0.9463       0.8211    205
Keyword ILIKE              0.7561       0.5821    205

Keyword vs Hybrid — Hit Rate delta: -0.1902
Keyword vs Hybrid — MRR delta:      -0.2390


### Field Weight Tuning

Demonstrate how changing `field_weights` affects keyword search results.
The default weights favour Question matches (2.0) over Answer matches (1.0).
Reversing them or adding Category weight changes the ranking.

In [25]:
weight_configs = [
    ("Question=1, Answer=1", {"Question": 1.0, "Answer": 1.0}),
    ("Question=2, Answer=1", {"Question": 2.0, "Answer": 1.0}),
    ("Question=1, Answer=2", {"Question": 1.0, "Answer": 2.0}),
    ("Question=3, Answer=1", {"Question": 3.0, "Answer": 1.0}),
    ("Question=1, Answer=3", {"Question": 1.0, "Answer": 3.0}),
    ("Question=1, Answer=1, Category=1", {"Question": 1.0, "Answer": 1.0, "Category": 1.0}),
]

print(f"{'Weights':<38} {'Hit Rate@5':>12} {'MRR@5':>12}")
print(f"{'-------':<38} {'----------':>12} {'------':>12}")
for label, weights in weight_configs:
    r = evaluate_retrieval(
        ground_truth=gt,
        k=5,
        search_fn=keyword_search,
        db_path=db_path,
        docs_path=docs_path,
        field_weights=weights,
    )
    print(f"{label:<38} {r['hit_rate']:>12.4f} {r['mrr']:>12.4f}")

Weights                                  Hit Rate@5        MRR@5
-------                                  ----------       ------
Question=1, Answer=1                         0.7512       0.5920
Question=2, Answer=1                         0.7561       0.5821
Question=1, Answer=2                         0.7415       0.5675
Question=3, Answer=1                         0.7268       0.5424
Question=1, Answer=3                         0.7024       0.5401
Question=1, Answer=1, Category=1             0.7512       0.5764


## 11. Retrieval Hyperparameter Tuning

Sweep the knobs that control retrieval quality — RRF K, k (top‑k cutoff), BM25 (k1, b), and
category weight for both hybrid and keyword search — and report which settings maximise
hit rate@k and MRR@k.

In [26]:
from functools import partial

from src.evaluate import evaluate_retrieval, load_ground_truth
from src.search import search

gt = load_ground_truth()

rrf_values = [1, 1.5, 2, 3, 5,10]
print(f"{'RRF_K':>6}  {'Hit Rate@5':>12}  {'MRR@5':>10}")
print(f"{'------':>6}  {'----------':>12}  {'------':>10}")
for rrf_k in rrf_values:
    r = evaluate_retrieval(
        ground_truth=gt, k=5,
        search_fn=partial(search, rrf_k=rrf_k),
        bm25_path=bm25_path, faiss_path=faiss_path, docs_path=docs_path,
    )
    print(f"{rrf_k:>6}  {r['hit_rate']:>12.4f}  {r['mrr']:>10.4f}")

 RRF_K    Hit Rate@5       MRR@5
------    ----------      ------
     1        0.9463      0.8260
   1.5        0.9463      0.8260
     2        0.9463      0.8252
     3        0.9463      0.8232
     5        0.9463      0.8187
    10        0.9463      0.8211


### k Sweep (Hybrid vs Keyword)

Evaluate how retrieval quality changes with different top‑k cutoffs for both hybrid and keyword search.

In [27]:
from src.search import keyword_search

k_values = [1, 3, 5, 10]
strategies = [
    ("Hybrid BM25+FAISS", lambda k: dict(
        search_fn=search,
        bm25_path=bm25_path, faiss_path=faiss_path, docs_path=docs_path,
    )),
    ("Keyword ILIKE", lambda k: dict(
        search_fn=keyword_search,
        db_path=db_path, docs_path=docs_path,
    )),
]

header = f"{'k':>3}"
for label, _ in strategies:
    header += f"  {label:^24}"
print(header)
header2 = f"{'':>3}"
for _ in strategies:
    header2 += f"  {'HitRate':>10}  {'MRR':>10}"
print(header2)
sep = f"{'':->3}"
for _ in strategies:
    sep += f"  {'':->10}  {'':->10}"
print(sep)

for k in k_values:
    line = f"{k:>3}"
    for _, kwargs_fn in strategies:
        r = evaluate_retrieval(ground_truth=gt, k=k, **kwargs_fn(k))
        line += f"  {r['hit_rate']:>10.4f}  {r['mrr']:>10.4f}"
    print(line)

  k     Hybrid BM25+FAISS           Keyword ILIKE      
        HitRate         MRR     HitRate         MRR
---  ----------  ----------  ----------  ----------
  1      0.6878      0.6878      0.4878      0.4878
  3      0.9171      0.8252      0.6390      0.5553
  5      0.9463      0.8211      0.7561      0.5821
 10      0.9805      0.8299      0.8878      0.5992


### BM25 k1/b Sweep

Sweep BM25Okapi parameters k1 and b, rebuild the index for each combination, and report the
best‑performing (k1, b) pair.

In [28]:
from src.faqs import load_faqs
from src.ingest import _combine_text, _tokenize, DEFAULT_BM25_PATH, DEFAULT_FAISS_PATH, DEFAULT_DOCS_PATH
from rank_bm25 import BM25Okapi
import pickle

k1_values = [0.5, 1.0, 1.5, 2.0]
b_values = [0.5, 0.75, 1.0]

faqs = load_faqs()
docs = [_combine_text(row) for row in faqs]
tokenized = [_tokenize(d) for d in docs]

print(f"{'k1':>5}  {'b':>5}  {'Hit Rate@5':>12}  {'MRR@5':>10}")
print(f"{'---':>5}  {'---':>5}  {'----------':>12}  {'------':>10}")

best = {"hit_rate": 0.0, "mrr": 0.0}
for k1 in k1_values:
    for b in b_values:
        bm25 = BM25Okapi(tokenized, k1=k1, b=b)
        with open(bm25_path, "wb") as f:
            pickle.dump(bm25, f)

        r = evaluate_retrieval(
            ground_truth=gt, k=5,
            bm25_path=bm25_path, faiss_path=faiss_path, docs_path=docs_path,
        )
        print(f"{k1:>5.1f}  {b:>5.2f}  {r['hit_rate']:>12.4f}  {r['mrr']:>10.4f}")

        if r['hit_rate'] > best['hit_rate'] or (r['hit_rate'] == best['hit_rate'] and r['mrr'] > best['mrr']):
            best = {"k1": k1, "b": b, "hit_rate": r['hit_rate'], "mrr": r['mrr']}

print()
print(f"Best: k1={best['k1']}, b={best['b']}  —  Hit Rate@5={best['hit_rate']:.4f}, MRR@5={best['mrr']:.4f}")

   k1      b    Hit Rate@5       MRR@5
  ---    ---    ----------      ------
  0.5   0.50        0.9415      0.8137
  0.5   0.75        0.9415      0.8177
  0.5   1.00        0.9366      0.8167
  1.0   0.50        0.9463      0.8169
  1.0   0.75        0.9415      0.8202
  1.0   1.00        0.9366      0.8196
  1.5   0.50        0.9512      0.8236
  1.5   0.75        0.9463      0.8211
  1.5   1.00        0.9463      0.8271
  2.0   0.50        0.9512      0.8260
  2.0   0.75        0.9561      0.8272
  2.0   1.00        0.9512      0.8287

Best: k1=2.0, b=0.75  —  Hit Rate@5=0.9561, MRR@5=0.8272


### Category Weight Sweep (Hybrid)

Tune how much the document's category field contributes to hybrid search ranking.
`cat_weight=0` means category plays no role (baseline). Higher values boost documents
whose category tokens appear in the query.

In [29]:
from functools import partial
from src.evaluate import evaluate_retrieval, load_ground_truth
from src.search import search

gt = load_ground_truth()

cat_weight_values = [0, 0.5, 1.0, 2.0, 3.0]
print(f"{'CatWeight':>10}  {'Hit Rate@5':>12}  {'MRR@5':>10}")
print(f"{'----------':>10}  {'----------':>12}  {'------':>10}")
for cat_weight in cat_weight_values:
    r = evaluate_retrieval(
        ground_truth=gt, k=5,
        search_fn=partial(search, cat_weight=cat_weight),
        bm25_path=bm25_path, faiss_path=faiss_path, docs_path=docs_path,
    )
    print(f"{cat_weight:>10.1f}  {r['hit_rate']:>12.4f}  {r['mrr']:>10.4f}")

 CatWeight    Hit Rate@5       MRR@5
----------    ----------      ------
       0.0        0.9512      0.8287
       0.5        0.9512      0.8064
       1.0        0.9512      0.8105
       2.0        0.9561      0.7813
       3.0        0.9024      0.7438


### Category Weight Sweep (Keyword)

Sweep the `Category` field weight in `keyword_search` with `Question=1, Answer=1` fixed.

In [30]:
from src.search import keyword_search
from src.evaluate import evaluate_retrieval, load_ground_truth

gt = load_ground_truth()

cat_values = [0, 0.5, 1.0, 2.0, 3.0]
print(f"{'Cat Weight':>10}  {'Hit Rate@5':>12}  {'MRR@5':>10}")
print(f"{'----------':>10}  {'----------':>12}  {'------':>10}")
for cat_w in cat_values:
    r = evaluate_retrieval(
        ground_truth=gt, k=5,
        search_fn=keyword_search,
        db_path=db_path, docs_path=docs_path,
        field_weights={"Question": 1.0, "Answer": 1.0, "Category": cat_w},
    )
    print(f"{cat_w:>10.1f}  {r['hit_rate']:>12.4f}  {r['mrr']:>10.4f}")

Cat Weight    Hit Rate@5       MRR@5
----------    ----------      ------
       0.0        0.7512      0.5920
       0.5        0.7512      0.5965
       1.0        0.7512      0.5764
       2.0        0.6976      0.5113
       3.0        0.6634      0.4668


## 12. Applying the tuned parameters

The §11 sweeps identified the following best retrieval settings on the
ground-truth set (`data/ground_truth.csv`):

| Parameter   | Chosen value | Why |
|-------------|--------------|-----|
| `k`         | 5            | k=10 raises hit rate 0.9463 → 0.9805 but roughly doubles LLM context/cost while MRR barely moves (0.8211 → 0.8299). |
| `rrf_k`     | 1            | Best MRR@5 (0.8260) vs 0.8252–0.8211 for 2–10; the previous code default of 60 was never tested. |
| `cat_weight`| 0            | Best MRR@5 (0.8287); the category-token boost does not help hybrid ranking. |
| `bm25_k1`   | 2.0          | With `b=0.75`: hit rate@5 0.9561, MRR@5 0.8272 (previous default was 1.5). |
| `bm25_b`    | 0.75         | Best `b` across the k1 sweep. |

These values are committed at the repo root in `tuned_params.json`. The application
code now reads them as production defaults:

- `src/ingest.py` `build_indexes()` uses `bm25_k1` / `bm25_b` for the BM25 index.
- `src/search.py` `search()` uses `k` / `rrf_k` / `cat_weight` for hybrid retrieval.
- `src/rag.py` `answer_question()` no longer hardcodes `k`; it inherits the
  config-driven default (an explicit `k` from callers still wins).

The cells below re-confirm the chosen settings with a short sweep, rebuild the local
BM25 index with the tuned `k1`/`b` so local runs match production, and verify the
committed config is actually wired into the code defaults.


In [ ]:
from functools import partial

from src.evaluate import evaluate_retrieval, load_ground_truth
from src.search import search

gt = load_ground_truth()

print("--- RRF_K ---")
for rrf_k in [1, 60]:
    r = evaluate_retrieval(
        ground_truth=gt, k=5,
        search_fn=partial(search, rrf_k=rrf_k),
        bm25_path=bm25_path, faiss_path=faiss_path, docs_path=docs_path,
    )
    print(f"RRF_K={rrf_k:>3}  Hit Rate@5={r['hit_rate']:.4f}  MRR@5={r['mrr']:.4f}")

print()
print("--- cat_weight ---")
for cat_weight in [0, 1.0]:
    r = evaluate_retrieval(
        ground_truth=gt, k=5,
        search_fn=partial(search, cat_weight=cat_weight),
        bm25_path=bm25_path, faiss_path=faiss_path, docs_path=docs_path,
    )
    print(f"cat_weight={cat_weight:>4.1f}  Hit Rate@5={r['hit_rate']:.4f}  MRR@5={r['mrr']:.4f}")

print()
print("--- BM25 k1/b ---")
from src.faqs import load_faqs
from src.ingest import _combine_text, _tokenize
from rank_bm25 import BM25Okapi
import pickle

faqs = load_faqs()
docs = [_combine_text(row) for row in faqs]
tokenized = [_tokenize(d) for d in docs]

for k1, b in [(2.0, 0.75), (1.5, 0.75)]:
    bm25 = BM25Okapi(tokenized, k1=k1, b=b)
    with open(bm25_path, "wb") as f:
        pickle.dump(bm25, f)
    r = evaluate_retrieval(
        ground_truth=gt, k=5,
        bm25_path=bm25_path, faiss_path=faiss_path, docs_path=docs_path,
    )
    print(f"k1={k1:.1f}  b={b:.2f}  Hit Rate@5={r['hit_rate']:.4f}  MRR@5={r['mrr']:.4f}")


In [ ]:
from src.config import load_tuned_params
from src.ingest import build_indexes

tuned = load_tuned_params()
paths = build_indexes(
    faqs=faqs,
    bm25_path=bm25_path,
    faiss_path=faiss_path,
    docs_path=docs_path,
    force=True,
    k1=tuned["bm25_k1"],
    b=tuned["bm25_b"],
)
print(f"Rebuilt indexes with tuned k1={tuned['bm25_k1']}, b={tuned['bm25_b']}")
for key, value in paths.items():
    print(f"  {key}: {value}")


In [ ]:
import json
import pathlib

from src.config import load_tuned_params

tuned_path = pathlib.Path("tuned_params.json")
with open(tuned_path, encoding="utf-8") as f:
    file_params = json.load(f)

defaults = load_tuned_params()
for key in ["k", "rrf_k", "cat_weight", "bm25_k1", "bm25_b"]:
    assert defaults[key] == file_params[key], f"{key}: code default {defaults[key]} != config {file_params[key]}"

print("Code defaults match tuned_params.json:")
for key in ["k", "rrf_k", "cat_weight", "bm25_k1", "bm25_b"]:
    print(f"  {key:>10} = {defaults[key]}")


## 13. LLM-as-a-judge Relevance Scoring

Evaluate RAG answer quality by asking an LLM to judge each response on two axes:
**relevance** (how well the answer addresses the question) and **faithfulness**
(whether the answer stays true to the retrieved context without hallucinating).

The `src.evaluate_llm` module wraps this into `evaluate_relevance()`, which runs
ground truth questions through the full RAG pipeline and then sends each
(question, answer, context) triplet to a judge LLM for composite scoring.

Cost tracking is built in: each row logs the RAG generation cost and the judge
LLM cost separately. The report sums both to give a full picture of evaluation
expenditure.

This section compares answer quality scores with the retrieval metrics from
section 9 — does better retrieval actually produce better answers?


In [31]:
import os

from src.evaluate import load_ground_truth
import random
from src.evaluate_llm import evaluate_relevance

has_keys = bool(os.environ.get("OPENAI_API_KEY"))

if has_keys:
    gt_sample = load_ground_truth()
    report = evaluate_relevance(
        ground_truth=gt_sample,
        bm25_path=bm25_path,
        faiss_path=faiss_path,
        docs_path=docs_path,
    )
    print(f"Queries evaluated: {report['n']}  |  Successfully judged: {report['n_valid']}")
    print()
    print(f"Mean Relevance:    {report['mean_relevance']}")
    print(f"Mean Faithfulness: {report['mean_faithfulness']}")
    print()
    print("--- Score Distribution ---")
    print(f"Relevance:    {report['distribution']['relevance']}")
    print(f"Faithfulness: {report['distribution']['faithfulness']}")
    print()
    print(f"--- Cost (${report['total_cost']:.6f}) ---")
    print(f"  RAG:   ${report['total_rag_cost']:.6f}  (${report['mean_rag_cost']:.6f}/query)")
    print(f"  Judge: ${report['total_judge_cost']:.6f}  (${report['mean_judge_cost']:.6f}/query)")
else:
    print("No API key found. Set OPENAI_API_KEY to run LLM-as-a-judge evaluation.")

Evaluating: 100%|██████████| 205/205 [16:22<00:00,  4.79s/it]

Queries evaluated: 205  |  Successfully judged: 205

Mean Relevance:    4.94
Mean Faithfulness: 5.0

--- Score Distribution ---
Relevance:    {1: 0, 2: 0, 3: 1, 4: 10, 5: 194}
Faithfulness: {1: 0, 2: 0, 3: 0, 4: 0, 5: 205}

--- Cost ($0.239010) ---
  RAG:   $0.099662  ($0.000486/query)
  Judge: $0.139348  ($0.000680/query)


In [33]:
if has_keys:
    from src.evaluate import evaluate_retrieval

    ret_report = evaluate_retrieval(
        ground_truth=gt_sample, k=5,
        bm25_path=bm25_path, faiss_path=faiss_path, docs_path=docs_path,
    )

    ret_by_q = {d["question"]: d for d in ret_report["details"]}

    print(f"{'Query':<48} {'Hit':>5} {'Rel':>5} {'Faith':>5}")
    print("-" * 48, f"{'---':>5} {'---':>5} {'----':>5}")
    for d in report["details"]:
        rd = ret_by_q.get(d["question"])
        hit = "✓" if rd and rd["hit"] else "✗" if rd else "?"
        rel = str(d["relevance"]) if d["relevance"] else "—"
        faith = str(d["faithfulness"]) if d["faithfulness"] else "—"
        print(f"{d["question"][:46]:<48} {hit:>5} {rel:>5} {faith:>5}")
    print()
    print(f"Retrieval hit rate@{ret_report["k"]}: {ret_report["hit_rate"]}")
    print(f"Mean relevance score:       {report["mean_relevance"]}")
    print(f"Mean faithfulness score:    {report["mean_faithfulness"]}")
    print()
    print(f"Evaluation cost: RAG=${report['total_rag_cost']:.6f} / Judge=${report['total_judge_cost']:.6f} / Total=${report['total_cost']:.6f}")


Query                                              Hit   Rel Faith
------------------------------------------------   ---   ---  ----
What payment options do you accept for renting       ✓     5     5
Can I pay with credit card, cash, or PayPal?         ✓     5     5
Do bounce house rentals take bank transfers to       ✓     5     5
Which payment methods are available when booki       ✓     5     5
Are there any fees for returned payments or bo       ✓     5     5
What’s your cancellation policy for a bounce h       ✓     5     5
How much do I get back if I cancel my bouncy c       ✗     4     5
If I cancel 2 weeks before, do I get a full re       ✓     5     5
What happens to my deposit if I cancel 5-13 da       ✓     5     5
How bad is the cancellation fee if I cancel le       ✓     5     5
How early should I book a bounce house rental?       ✓     5     5
Can I reserve a bouncy castle months ahead?          ✗     3     5
How far ahead do you take bookings?                  ✓     5  